src/ML/pytorch/valueMLmodel.cc

```diff
@@ -0,0 +1,84 @@
+    // create a multi-threaded data loader for the MNIST dataset.
+    auto data_loader = torch::data::make_data_loader<torch::data::samplers::SequentialSampler>(std::move(data_set), batch_size);
+    // instantiate an SGD optimization algorithm to update our LinearNet's parameters.
+    torch::optim::SGD optimizer(linearNet->parameters(), /*lr=*/alpha);
+    for (size_t epoch = 1; epoch <= numberOfEpochs; ++epoch) {
+        torch::Tensor loss;
+        // Iterate the data loader to yield batches from the dataset.
+        for (auto& batch : *data_loader) {
+            // Reset gradients.
+            optimizer.zero_grad();
+            // Execute the model on the input data.
+            torch::Tensor prediction = linearNet->forward(batch.data);
+            // Compute a loss value to judge the prediction of our model.
+            loss = torch::mse_loss(prediction, batch.target);
+            // Compute gradients of the loss w.r.t. the parameters of our model.
+            loss.backward();
+            // Update the parameters based on the calculated gradients.
+            optimizer.step();
+            // Output the loss and checkpoint every 100 batches
+        }
+        std::cout << "Epoch: " << epoch << " | Loss: " << loss.item<float>() << std::endl;
+    }
+}
+
+vector<float> valueMLmodel::calcValueEstimates(state* s) {
+    /* Calculate all possible successor states of s.
+       Returns a boolean vector corresponding to edges in such a way that true means an edge is selectable */
+    vector<bool> index = s->calcSuccessorStates();     
+
+    torch::Tensor succStates = vectorToTensor(s->successorStates);
+
+    // Returns the value prediction for all successor states.       
+    torch::Tensor pred = linearNet->forward(succStates);
+
+    vector<float> tempPrediction = tensorToVector(pred);
+    vector<float> prediction;
+    unsigned int counter = 0;
+
+    // To get a vector of the size of the state, we set 0 for every not selectable edge.
+    for (unsigned int i = 0; i < dimstate; i++) {
+        if (index.at(i)) {
+            prediction.push_back(tempPrediction.at(counter));
+            counter++;
+        } else {
+        prediction.push_back(0);
+        }
+    }                                        
+    return prediction;                              
+}
+
+// Transforms a vector of states into a tensor
+torch::Tensor valueMLmodel::vectorToTensor(vector<vector<bool>>& vec) {
+    torch::Tensor tens = torch::zeros({(long int) vec.size(),(long int) vec.at(0).size()});
+    for (unsigned int i = 0; i < vec.size(); i++){
+        for (unsigned int j = 0; j < vec.at(i).size(); j++) {                
+            tens[i][j] = (float)vec.at(i).at(j);
+        }           
+    }
+    return tens;
+}
+
+// Transforms a tensor of values for the successor states into a vector
+vector<float> valueMLmodel::tensorToVector(torch::Tensor& tens) {
+    vector<float*> vec;
```

Comment:
Please clean up: `vec` is not needed. Also, please initialize vecReturn with size and then do not use push_back.

Reply:
We're hopefully completely removing both functions and just working with tensors after some talks with Felix.